packages/astro/src/core/app/ssrPipeline.ts

```diff
@@ -0,0 +1,54 @@
+import type { Environment } from '../render';
+import type { EndpointCallResult } from '../endpoint/index.js';
+import mime from 'mime';
+import { attachCookiesToResponse } from '../cookies/index.js';
+import { Pipeline } from '../pipeline.js';
+
+/**
+ * Thrown when an endpoint contains a response with the header "X-Astro-Response" === 'Not-Found'
+ */
+export class EndpointNotFoundError extends Error {
+	originalResponse: Response;
+	constructor(originalResponse: Response) {
+		super();
+		this.originalResponse = originalResponse;
+	}
+}
+
+export class SSRRoutePipeline extends Pipeline {
+	encoder = new TextEncoder();
+
+	constructor(env: Environment) {
+		super(env);
+		this.setEndpointHandler(this.#ssrEndpointHandler);
+	}
+
+	// This function is responsible for handling the result coming from an endpoint.
+	async #ssrEndpointHandler(request: Request, response: EndpointCallResult): Promise<Response> {
```

Comment:
Small nit: Is `this.setEndpointHandler` and `this.#ssrEndpointHandler` something that every extended class of `Pipeline` needs to implement? I was thinking we can create an abstract method like `abstract endpointHandler()` in `Pipeline` instead, but I'm also fine with this pattern if you're still shaping up the final pattern.

Reply:
Technically, no.

Not all the future pipelines will have to handle endpoints. For example: a testing pipeline will need to be used only when needed, an Astro container won't need it (I presume)